R/utils.R

```diff
@@ -803,3 +803,179 @@ doPrunePed <- function(ped, addMissingMates=FALSE, solveMultiGraph="use.all"){
+#' mbped <- mbsub[, c("famid", "id", "fatherid", "motherid", "sex")]
+#' ## Renaming column names
+#' colnames(mbped) <- c("family", "id", "father", "mother", "sex")
+#' ## Defining the optional argument age.
+#' Age <- mbsub$endage
+#' names(Age) <- mbsub$id
+#' ## Create the object
+#' fad <- FAData(pedigree=mbped, age=Age)
+#'
+#' ## Getting all pairs of individuals with a kinship coefficient >= 0.25
+#' ## keeping all duplicates
+#' rel_pairs <- kinshipPairs(fad)
+#' head(rel_pairs)
+#' ## As we see, we have multiple times the individual 1 etc.
+#'
+#' ## For an actual correlation analysis it would be better to drop duplicates.
+#' ## Below we randomly select individual pairs if they occurr multiple times
+#' rel_pairs <- kinshipPairs(fad, duplicates = "random")
+#' head(rel_pairs)
+#'
+#' ## In addition we extract pairs of individuals that are much less related.
+#' ## For this examples we consider all individuals with a kinship
+#' ## coefficient < 0.03125 (second cousin) to be *unrelated*.
+#' unrel_pairs <- kinshipPairs(fad, duplicates = "random",
+#'     condition = function(z) z < 0.03125)
+#' head(unrel_pairs)
+#'
+#' ## For a familial resemblance analysis we can now calculate the correlation
+#' ## coefficient of a quantitative trait between pairs of related individuals
+#' ## and compare that with the correlation coefficient calculated on unrelated
+#' ## individuals. For our toy example we use the participant's age, since we
+#' ## don't have any other quantitative values available.
+#' cor_rel <- cor(age(fad)[rel_pairs[, 1]], age(fad)[rel_pairs[, 2]],
+#'     use = "pairwise.complete.obs")
+#'
+#' cor_unrel <- cor(age(fad)[unrel_pairs[, 1]], age(fad)[unrel_pairs[, 2]],
+#'     use = "pairwise.complete.obs")
+#' cor_rel
+#' cor_unrel
+#'
+#' ## We don't see a clear difference in the correlation, thus, the age (as
+#' ## expected) age has no familial component.
+kinshipPairs <- function(x, condition = function(x) x >= 0.25,
+                         duplicates = c("keep", "first", "last", "random"),
+                         id = NULL, family = NULL) {
+    duplicates <- match.arg(duplicates)
+    if (!inherits(x, "FAData"))
+        stop("'kinshipPairs' is currently only implemented for 'FAData'")
+    if (length(family)) {
+        id <- x$id[x$family %in% family]
+        if (!length(id))
+            stop("No individuals for the provided families found")
+    }
+    ks <- kinship(x)
+    if (length(id)) {
+        is_id <- rownames(ks) %in% id
+        ks <- ks[is_id, is_id]
+    }
+    res <- condition(ks)
+    if (!length(dim(res)) || ncol(ks) != ncol(res) ||
+        nrow(ks) != nrow(res) || !is.logical(res[1, 1]))
+        stop("'condition' function did not return a 'logical' 'matrix' with ",
+             "dimensions matching the kinship matrix in 'x'")
+    res[lower.tri(res, diag = TRUE)] <- NA
+    if (!length(dim(res)) || nrow(res) < 1)
+        return(matrix(nrow = 0, ncol = 2))
```

Comment:
Any specific reason to return a `matrix`, especially as the rows always contain `character` data? Is it better for indexing the kinship matrix?

Reply:
the function should return *pairs of individuals* thus I believe the ID of the individual is better suited here than the index of the individual in the kinship matrix (which can be extremely misleading if the ID of an individual is a simple number). Thus I'm returning always a `character` `matrix` to not confuse the user (is it an index? is it the ID?)

Reply:
For me this would have been naturally a data frame. Matrices occopy the numerical part (at least in my head). With the data frame one could move on more naturally in a workflow, e.g. split by ID, merge with other data (frames), etc.

Reply:
we could have it either way. Since it's just one data type (`characer`) and only two columns a `matrix` also makes sense (note that `matrix` is not only a numeric matrix! it's a 2 dimensional array with a single data type). subsetting/accessing rows of a `matrix` is generally faster than for a `data.frame`. Another thing is that we should consider each row as being a *pair* - that's more implicit in a `matrix` than in a `data.frame` (since the different columns of a `data.frame` can be a different data type.

Also note that within the `for` loop inside the function I'm adding always the pair to one row in the result array. Doing this with a `data.frame` is totally inefficient (because in fact a `data.frame` is simply a `list` with the constraint that each list elelemt needs to have the same length). So, replacing a *row* means you have to replace one element in the first column and then an element in the second column. Thus in `R`, if you work on rows and want to be computationally more efficient, don't use `data.frame`.

If I did not convince you I can change to a `data.frame` - but this would mean to call `as.data.frame` on the resulting result `matrix` :)

Reply:
Well, this can then be done by the caller.